From 4195776df2a7d19e2381e86a7e5bee3411edc205 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Thu, 1 Jun 2023 17:25:41 +0800
Subject: [PATCH 396/780] [Huawei] blk-io-hierarchy: main data structure
 definition

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188826

--------------------------------

Add definitions of struct blk_io_hierarchy_stats and add it to
request_queue.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-io-hierarchy/stats.h | 34 ++++++++++++++++++++++++++++++++++
 include/linux/blk_types.h      |  5 +++++
 include/linux/blkdev.h         |  4 ++++
 3 files changed, 43 insertions(+)
 create mode 100644 block/blk-io-hierarchy/stats.h

diff --git a/block/blk-io-hierarchy/stats.h b/block/blk-io-hierarchy/stats.h
new file mode 100644
index 000000000000..87ee80152d22
--- /dev/null
+++ b/block/blk-io-hierarchy/stats.h
@@ -0,0 +1,34 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2023. Huawei Technologies Co., Ltd. All rights reserved.
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef BLK_IO_HIERARCHY_STATS_H
+#define BLK_IO_HIERARCHY_STATS_H
+
+#ifdef CONFIG_BLK_IO_HIERARCHY_STATS
+
+#include <linux/blkdev.h>
+
+struct hierarchy_stats {
+	u64 nsecs[NR_STAGE_GROUPS][NR_STAT_GROUPS];
+	unsigned long dispatched[NR_STAGE_GROUPS][NR_STAT_GROUPS];
+	unsigned long completed[NR_STAGE_GROUPS][NR_STAT_GROUPS];
+};
+
+struct blk_io_hierarchy_stats {
+	struct request_queue *q;
+	struct dentry *debugfs_dir;
+	struct hierarchy_stats __percpu *stats;
+};
+
+#endif /* CONFIG_BLK_IO_HIERARCHY_STATS */
+#endif /* BLK_IO_HIERARCHY_STATS_H */
diff --git a/include/linux/blk_types.h b/include/linux/blk_types.h
index c4e17f180669..e900de98982f 100644
--- a/include/linux/blk_types.h
+++ b/include/linux/blk_types.h
@@ -464,6 +464,11 @@ enum stat_group {
 	NR_STAT_GROUPS
 };
 
+enum stage_group {
+	STAGE_RESERVE = 0,
+	NR_STAGE_GROUPS,
+};
+
 #define bio_op(bio) \
 	((bio)->bi_opf & REQ_OP_MASK)
 #define req_op(req) \
diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index 1c8cc129cc5c..6a45fa5fcc8a 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -603,6 +603,10 @@ struct request_queue {
 #define BLK_MAX_WRITE_HINTS	5
 	u64			write_hints[BLK_MAX_WRITE_HINTS];
 
+#ifdef CONFIG_BLK_IO_HIERARCHY_STATS
+	struct blk_io_hierarchy_stats *io_hierarchy_stats;
+#endif
+
 	KABI_REPLACE(unsigned long dtag_wait_time,
 		     struct blk_mq_tags *shared_sbitmap_tags)
 	KABI_RESERVE(1)
-- 
2.17.1

